crates/bevy_ecs/src/template.rs

```diff
@@ -0,0 +1,190 @@
+};
+use alloc::{boxed::Box, vec, vec::Vec};
+use bevy_platform::collections::hash_map::Entry;
+use bevy_utils::TypeIdMap;
+use core::any::{Any, TypeId};
+use downcast_rs::{impl_downcast, Downcast};
+use variadics_please::all_tuples;
+
+/// A [`Template`] is something that, given a spawn context (target [`Entity`], [`World`](crate::world::World), etc), can produce a [`Template::Output`].
+pub trait Template {
+    /// The type of value produced by this [`Template`].
+    type Output;
+
+    /// Uses this template and the given `entity` context to produce a [`Template::Output`].
+    fn build(&mut self, entity: &mut EntityWorldMut) -> Result<Self::Output>;
+
+    /// This is used to register information about the template, such as dependencies that should be loaded before it is instantiated.
+    #[inline]
+    fn register_data(&self, _data: &mut TemplateData) {}
+}
+
+/// [`GetTemplate`] is implemented for types that can be produced by a specific, canonical [`Template`]. This creates a way to correlate to the [`Template`] using the
+/// desired template output type. This is used by Bevy's scene system.
+pub trait GetTemplate: Sized {
+    /// The [`Template`] for this type.
+    type Template: Template;
+}
+
+macro_rules! template_impl {
+    ($($template: ident),*) => {
+        #[expect(
+            clippy::allow_attributes,
+            reason = "This is a tuple-related macro; as such, the lints below may not always apply."
+        )]
+        impl<$($template: Template),*> Template for TemplateTuple<($($template,)*)> {
+            type Output = ($($template::Output,)*);
+            fn build(&mut self, _entity: &mut EntityWorldMut) -> Result<Self::Output> {
+                #[allow(
+                    non_snake_case,
+                    reason = "The names of these variables are provided by the caller, not by us."
+                )]
+                let ($($template,)*) = &mut self.0;
+                Ok(($($template.build(_entity)?,)*))
+            }
+
+            fn register_data(&self, _data: &mut TemplateData) {
+                #[allow(
+                    non_snake_case,
+                    reason = "The names of these variables are provided by the caller, not by us."
+                )]
+                let ($($template,)*) = &self.0;
+                $($template.register_data(_data);)*
+            }
+        }
+    }
+}
+
+/// A wrapper over a tuple of [`Template`] implementations, which also implements [`Template`]. This exists because [`Template`] cannot
+/// be directly implemented for tuples of [`Template`] implementations.
+pub struct TemplateTuple<T>(pub T);
+
+all_tuples!(template_impl, 0, 12, T);
+
+impl Template for EntityPath<'static> {
+    type Output = Entity;
+
+    fn build(&mut self, entity: &mut EntityWorldMut) -> Result<Self::Output> {
+        Ok(entity.resolve_path(self)?)
+    }
+}
+
+impl GetTemplate for Entity {
+    type Template = EntityPath<'static>;
+}
+
+impl<T: Clone + Default> Template for T {
```

Comment:
The `Default` bound doesn't seem to actually be used here.  Are you expecting to use `T::default()` for something, or was this just to avoid conflicts with other impls?  And, in either case, does it make sense to generalize it to `FromWorld`?

Reply:
See also my [comment above] about "do we still even need FromWorld" too.